tests/test_format_detection.py

```diff
@@ -0,0 +1,169 @@
+from unittest import TestCase
+import os.path as osp
+
+from datumaro.components.format_detection import (
+    FormatDetectionConfidence, FormatRequirementsUnmet, apply_format_detector,
+)
+from datumaro.util.test_utils import TestDir
+
+from tests.requirements import Requirements, mark_requirement
+
+
+class FormatDetectionTest(TestCase):
+    def setUp(self) -> None:
+        test_dir_context = TestDir()
+        self._dataset_root = test_dir_context.__enter__()
+        self.addCleanup(test_dir_context.__exit__, None, None, None)
+
+    @mark_requirement(Requirements.DATUM_GENERAL_REQ)
+    def test_empty_detector(self):
+        result = apply_format_detector(self._dataset_root, lambda c: None)
+        self.assertEqual(result, FormatDetectionConfidence.MEDIUM)
+
+    @mark_requirement(Requirements.DATUM_GENERAL_REQ)
+    def test_custom_confidence(self):
+        result = apply_format_detector(self._dataset_root,
+            lambda c: FormatDetectionConfidence.LOW)
+        self.assertEqual(result, FormatDetectionConfidence.LOW)
+
+    @mark_requirement(Requirements.DATUM_GENERAL_REQ)
+    def test_root_path(self):
+        provided_root: str
+        def detect(context):
+            nonlocal provided_root
+            provided_root = context.root_path
```

Comment:
Consider using a mock.

Reply:
You mean as the detector? I don't see a great way to check the `context.root_path` attribute if I do that. It would be possible to retrieve the context from the mock _after_ the call to `apply_format_detector` (and then get `root_path` from it), but the format detection API implies that the context should not be accessed outside the detector function, and I would rather not have the tests violate that.

Reply:
I thought you could use `.assert_called_*` and check input args, but probably you're right about inaccessible object.

Reply:
> I thought you could use .assert_called_* and check input args

Yeah, that would definitely not work - the argument is a context object, which is not comparable.

I was thinking more along the lines of `self.assertEqual(detector.call_args[0].root_path, self._dataset_root)`. But, to reiterate, I don't like that this smuggles the context out of the detector, since by the semantics it's only supposed to be valid while in the detector.